libs/gl-plugin/src/tramp.rs

```diff
@@ -0,0 +1,354 @@
+            )
+            .await
+        });
+        part_id += 1;
+        handles.push(handle);
+    }
+
+    let results = join_all(handles).await;
+    let mut payment_preimage = None;
+    for result in results {
+        let response = result??;
+        if let Some(preimage) = response.payment_preimage {
+            payment_preimage = Some(preimage);
+        }
+    }
+
+    if let Some(payment_preimage) = payment_preimage {
+        Ok(cln_rpc::model::responses::PayResponse {
+            destination: Some(decoded.payee),
+            warning_partial_completion: None,
+            status: cln_rpc::model::responses::PayStatus::COMPLETE,
+            amount_msat: cln_rpc::primitives::Amount::from_msat(amount_msat),
+            amount_sent_msat: cln_rpc::primitives::Amount::from_msat(amount_msat),
+            created_at: 0.,
+            parts: parts as u32,
+            payment_hash: decoded.payment_hash,
+            payment_preimage,
+        })
+    } else {
+        Err(anyhow!("missing payment_preimage"))
+    }
+}
+
+async fn do_pay(
+    rpc: &mut ClnRpc,
+    node_id: PublicKey,
+    bolt11: String,
+    label: String,
+    description: Option<String>,
+    part_amt: u64,
+    scid: ShortChannelId,
+    part_id: u64,
+    group_id: u64,
+    payment_hash: cln_rpc::primitives::Sha256,
+    total_amount: Amount,
+    payment_secret: cln_rpc::primitives::Secret,
+    payment_metadata: String,
+    max_delay: Option<u32>,
+) -> Result<cln_rpc::model::responses::WaitsendpayResponse> {
+    let route = cln_rpc::model::requests::SendpayRoute {
+        amount_msat: cln_rpc::primitives::Amount::from_msat(part_amt),
+        id: node_id.clone(),
+        delay: max_delay.unwrap_or(MAX_DELAY_DEFAULT) as u16,
+        channel: scid,
+    };
+
+    debug!(
+        "Trampoline payment part_id={} with amount={}, using route={:?}",
+        part_id, part_amt, route
+    );
+
+    let _r: serde_json::Value = rpc
+        .call_raw(
+            "sendpay",
+            &SendpayRequest {
+                route: vec![route],
+                payment_hash,
+                label: as_option(label),
+                amount_msat: Some(total_amount),
+                bolt11: Some(bolt11),
+                payment_secret: Some(payment_secret),
+                partid: Some(part_id),
+                localinvreqid: None,
+                groupid: Some(group_id),
+                description,
+                payment_metadata: Some(payment_metadata),
+            },
+        )
+        .await?;
+
+    match rpc
+        .call_typed(&cln_rpc::model::requests::WaitsendpayRequest {
+            payment_hash: payment_hash,
+            timeout: Some(120),
+            partid: Some(part_id),
+            groupid: Some(group_id),
+        })
+        .await
+    {
+        Ok(r) => Ok(r),
+        Err(e) => Err(anyhow!(e.to_string())),
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct SendpayRequest {
```

Comment:
I'm surprised this is necessary, shouldn't we have a Pay request like this already in cln-rpc?